src/main/java/eu/interop/federationgateway/validator/DiagnosisKeyBatchValidator.java

```diff
@@ -15,24 +17,59 @@
   @Override
   public boolean isValid(EfgsProto.DiagnosisKeyBatch diagnosisKeyBatch, ConstraintValidatorContext context) {
 
+    /*
+     see
+       - https://github.com/google/exposure-notifications-server/blob/5a77e6982b16c7282c77dea39772491cc7b7dc8b/internal/publish/model/exposure_model.go#L338
+       - https://github.com/google/exposure-notifications-server/blob/5a77e6982b16c7282c77dea39772491cc7b7dc8b/tools/export-analyzer/main.go#L123
+     */
+
+    long minimumRollingStart = Instant
+      .now()
+      .truncatedTo(ChronoUnit.DAYS)
+      .minus(15, ChronoUnit.DAYS)
+      .getEpochSecond() / 600;
+
+    long maximumRollingStart = Instant
+      .now()
+      .getEpochSecond() / 600;
+    maximumRollingStart += 1;
+
     List<EfgsProto.DiagnosisKey> diagnosisKeys = diagnosisKeyBatch.getKeysList();
     for (EfgsProto.DiagnosisKey diagnosisKey : diagnosisKeys) {
-      if (diagnosisKey.getReportType() == EfgsProto.ReportType.UNKNOWN) {
-        log.error(VALIDATION_FAILED_MESSAGE + "Invalid report-type.");
-        return false;
-      } else if (diagnosisKey.getKeyData() == null || diagnosisKey.getKeyData().isEmpty()) {
-        log.error(VALIDATION_FAILED_MESSAGE + "The keydata is empty or null.");
-        return false;
-      } else if (diagnosisKey.getRollingPeriod() == 0) {
-        log.error(VALIDATION_FAILED_MESSAGE + "Invalid rolling period.");
-        return false;
-      } else if (diagnosisKey.getTransmissionRiskLevel() == 0 || diagnosisKey.getTransmissionRiskLevel() > 8) {
-        log.error(VALIDATION_FAILED_MESSAGE + "Invalid transmission risk level.");
-        return false;
-      }
+      if (diagnosisKey.getKeyData() == null || diagnosisKey.getKeyData().isEmpty()) {
+        return fail("The keydata is empty or null.", context);
+
+      } else if (diagnosisKey.getKeyData().size() != 16) {
+        return fail("The keydata is not 16 bytes.", context);
+
+      } else if (diagnosisKey.getRollingStartIntervalNumber() < minimumRollingStart
+        || diagnosisKey.getRollingStartIntervalNumber() > maximumRollingStart) {
+        return fail("Invalid rolling start interval number.", context);
+      } else if (diagnosisKey.getRollingPeriod() < 1 || diagnosisKey.getRollingPeriod() > 144) {
+        return fail("Invalid rolling period.", context);
+
+      } else if ((diagnosisKey.getTransmissionRiskLevel() < 0 || diagnosisKey.getTransmissionRiskLevel() > 8)
+        && diagnosisKey.getTransmissionRiskLevel() != 0x7fffffff) {
```

Comment:
0x7fffffff should be defined as a named constant - just looks like a magic number as is.

I still recommend against using this value - it's not a documented part of the export.proto as defined by GAEN

Reply:
Agree with you. I will change this when @SchulzeS decides to let the 0x7fffffff value allowed.

Reply:
0xf7777777 is replaced by a constant